File: lessons/compiled_lang/cpp/src/o00.flags.cpp
#include <iostream>
#include <ostream>
#include <string>

// usage() prints to whichever stream the caller passes - std::cout for an
//  explicit -h/-?, std::cerr for a usage error - matching the bash
//  reference's single usage() body, invoked either bare or with ">&2".
void usage(std::ostream& os, const char* cmd) {
    os << std::endl;
    os << "Usage: " << cmd << " [-c|-e|-l|-k|-p|-m|-t] [-h|-?]" << std::endl;
    os << std::endl;
    os << "  -c  Coffee" << std::endl;
    os << "  -e  Espresso" << std::endl;
    os << "  -l  Latte" << std::endl;
    os << "  -k  Machiato" << std::endl;
    os << "  -p  Capucino" << std::endl;
    os << "  -m  Mocha" << std::endl;
    os << "  -t  Tea" << std::endl;
    os << "  -h  Display this help message" << std::endl;
    os << "  -?  Display this help message" << std::endl;
    os << std::endl;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        usage(std::cerr, argv[0]);
        return 1;
    }

    std::string flag = argv[1];
    if (flag == "-c") { std::cout << "You ordered a Coffee." << std::endl; return 0; }
    if (flag == "-e") { std::cout << "You ordered an Espresso." << std::endl; return 0; }
    if (flag == "-l") { std::cout << "You ordered a Latte." << std::endl; return 0; }
    if (flag == "-k") { std::cout << "You ordered a Machiato." << std::endl; return 0; }
    if (flag == "-p") { std::cout << "You ordered a Capucino." << std::endl; return 0; }
    if (flag == "-m") { std::cout << "You ordered a Mocha." << std::endl; return 0; }
    if (flag == "-t") { std::cout << "You ordered a Tea." << std::endl; return 0; }
    if (flag == "-h" || flag == "-?") {
        usage(std::cout, argv[0]);
        return 0;
    }

    usage(std::cerr, argv[0]);
    return 1;
}

File: lessons/win_scripts/wsh.jscript/scripts/o00.flags.js
var usage = "\n" +
            "Usage: " + WScript.ScriptName + " [-c|-e|-l|-k|-p|-m|-t] [-h|-?]\n" +
            "\n" +
            "  -c  Coffee\n" +
            "  -e  Espresso\n" +
            "  -l  Latte\n" +
            "  -k  Machiato\n" +
            "  -p  Capucino\n" +
            "  -m  Mocha\n" +
            "  -t  Tea\n" +
            "  -h  Display this help message\n" +
            "  -?  Display this help message\n" +
            "\n";

if (WScript.Arguments.length == 1) {
   var flag = WScript.Arguments(0);
   switch (flag) {
      case "-c": WScript.Echo("You ordered a Coffee."); WScript.Quit(0);
      case "-e": WScript.Echo("You ordered an Espresso."); WScript.Quit(0);
      case "-l": WScript.Echo("You ordered a Latte."); WScript.Quit(0);
      case "-k": WScript.Echo("You ordered a Machiato."); WScript.Quit(0);
      case "-p": WScript.Echo("You ordered a Capucino."); WScript.Quit(0);
      case "-m": WScript.Echo("You ordered a Mocha."); WScript.Quit(0);
      case "-t": WScript.Echo("You ordered a Tea."); WScript.Quit(0);
      case "-h":
      case "-?":
         WScript.stdout.write(usage);
         WScript.Quit(0);
   }
}

WScript.stderr.write(usage);
WScript.Quit(1);
